src/Content/OEmbed.php

```diff
@@ -0,0 +1,352 @@
+	 * and replace it with [embed]url[/embed]
+	 */
+	public static function HTML2BBCode($text)
+	{
+		// start parser only if 'oembed' is in text
+		if (strpos($text, "oembed")) {
+
+			// convert non ascii chars to html entities
+			$html_text = mb_convert_encoding($text, 'HTML-ENTITIES', mb_detect_encoding($text));
+
+			// If it doesn't parse at all, just return the text.
+			$dom = @DOMDocument::loadHTML($html_text);
+			if (!$dom) {
+				return $text;
+			}
+			$xpath = new DOMXPath($dom);
+
+			$xattr = OEmbed::buildXPath("class", "oembed");
+			$entries = $xpath->query("//span[$xattr]");
+
+			$xattr = "@rel='oembed'"; //oe_build_xpath("rel","oembed");
+			foreach ($entries as $e) {
+				$href = $xpath->evaluate("a[$xattr]/@href", $e)->item(0)->nodeValue;
+				if (!is_null($href)) {
+					$e->parentNode->replaceChild(new DOMText("[embed]" . $href . "[/embed]"), $e);
+				}
+			}
+			return OEmbed::getInnerHTML($dom->getElementsByTagName("body")->item(0));
+		} else {
+			return $text;
+		}
+	}
+
+	/**
+	 * @brief Generates the iframe HTML for an oembed attachment.
+	 *
+	 * Width and height are given by the remote, and are regularly too small for
+	 * the generated iframe.
+	 *
+	 * The width is entirely discarded for the actual width of the post, while fixed
+	 * height is used as a starting point before the inevitable resizing.
+	 *
+	 * Since the iframe is automatically resized on load, there are no need for ugly
+	 * and impractical scrollbars.
+	 *
+	 * @param string $src Original remote URL to embed
+	 * @param string $width
+	 * @param string $height
+	 * @return string formatted HTML
+	 *
+	 * @see oembed_format_object()
+	 */
+	private static function iframe($src, $width, $height)
+	{
+		$a = get_app();
+
+		if (!$height || strstr($height, '%')) {
+			$height = '200';
+		}
+		$width = '100%';
+
+		$s = System::baseUrl() . '/oembed/' . base64url_encode($src);
+		return '<iframe onload="resizeIframe(this);" class="embed_rich" height="' . $height . '" width="' . $width . '" src="' . $s . '" allowfullscreen scrolling="no" frameborder="no">' . t('Embedded content') . '</iframe>';
+	}
+
+	/**
+	 * Generates an XPath query to select elements whose provided attribute contains
+	 * the provided value in a space-separated list.
+	 *
+	 * @brief Generates attribute search XPath string
+	 *
+	 * @param string $attr Name of the attribute to seach
+	 * @param string $value Value to search in a space-separated list
+	 * @return string
+	 */
+	private static function buildXPath($attr, $value)
+	{
+		// https://www.westhoffswelt.de/blog/2009/6/9/select-html-elements-with-more-than-one-css-class-using-xpath
+		return "contains( normalize-space( @$attr ), ' $value ' ) or substring( normalize-space( @$attr ), 1, string-length( '$value' ) + 1 ) = '$value ' or substring( normalize-space( @$attr ), string-length( @$attr ) - string-length( '$value' ) ) = ' $value' or @$attr = '$value'";

```

Comment:
There are several spaces in that line that could be removed to make it look better.

Reply:
Some spaces are required are it is looking for a value in a space-separated list, but there are a several others I can indeed remove.